internal/client/client.go

```diff
@@ -0,0 +1,419 @@
+		return v1types.ECOSYSTEM_GO
+	case strings.HasPrefix(purl, "pkg:npm"):
+		return v1types.ECOSYSTEM_NPM
+	case strings.HasPrefix(purl, "pkg:pypi"):
+		return v1types.ECOSYSTEM_PYPI
+	default:
+		return v1types.Ecosystem(0)
+	}
+}
+
+// PurlToDependency takes a string with a package url
+func (t *Trusty) PurlToDependency(purlString string) (*v1types.Dependency, error) {
+	e := t.PurlToEcosystem(purlString)
+	if e == 0 {
+		// Ecosystem nil or not supported
+		return nil, fmt.Errorf("ecosystem not supported")
+	}
+
+	purl, err := packageurl.FromString(purlString)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse package url: %w", err)
+	}
+	name := purl.Name
+	if purl.Namespace != "" {
+		name = purl.Namespace + "/" + purl.Name
+	}
+	return &v1types.Dependency{
+		Ecosystem: e,
+		Name:      name,
+		Version:   purl.Version,
+	}, nil
+}
+
+// Report returns a dependency report with all the data that Trusty has
+// available for a package.
+func (t *Trusty) Report(_ context.Context, dep *v1types.Dependency) (*v1types.Reply, error) {
+	u, err := t.PackageEndpoint(dep)
+	if err != nil {
+		return nil, fmt.Errorf("computing package endpoint: %w", err)
+	}
+
+	var r v1types.Reply
+	tries := 0
+	for {
+		resp, err := t.Options.HttpClient.GetRequest(u)
+		if err != nil {
+			return nil, fmt.Errorf("could not send request: %w", err)
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("received non-200 response: %d", resp.StatusCode)
+		}
+
+		dec := json.NewDecoder(resp.Body)
+		if err := dec.Decode(&r); err != nil {
+			return nil, fmt.Errorf("could not unmarshal response: %w", err)
+		}
+		fmt.Printf("Attempt #%d to fetch package, status: %s", tries, r.PackageData.Status)
+
+		shouldRetry, err := evalRetry(r.PackageData.Status, t.Options)
+		if err != nil {
+			return nil, err
+		}
+
+		if !shouldRetry {
+			break
+		}
+
+		tries++
+		if tries > t.Options.IngestionMaxRetries {
+			return nil, fmt.Errorf("time out reached waiting for package ingestion")
+		}
+		time.Sleep(time.Duration(t.Options.IngestionRetryWait) * time.Second)
+	}
+
+	return &r, err
+}
+
+func evalRetry(status string, opts Options) (shouldRetry bool, err error) {
+	// First, error if the ingestion status is invalid
+	if status != v1types.IngestStatusFailed && status != v1types.IngestStatusComplete &&
+		status != v1types.IngestStatusPending && status != v1types.IngestStatusScoring {
+
+		return false, fmt.Errorf("unexpected ingestion status when querying package")
+	}
+
+	if status == v1types.IngestStatusFailed && opts.ErrOnFailedIngestion {
+		return false, fmt.Errorf("upstream error ingesting package data")
+	}
+
+	// Package ingestion is ready
+	if status == v1types.IngestStatusComplete {
+		return false, nil
+	}
+
+	// Client configured to return raw response (even when package is not ready)
+	if !opts.WaitForIngestion || status == v1types.IngestStatusFailed {
+		return false, nil
+	}
+
+	return true, nil
+}
+
```

Comment:
Note to the reviewer: below this line there are the relevant changes, while all of the above was just moved from a different package.